Revert #126024 "Do not use global caches if opaque types can be defined"

lqd opened the pull request:
This PR prepares a revert of #126024 as discussed on zulip, because of its regressions to compilation times vs unclear soundness fix. Opening as draft for discussion between @compiler-errors and @lcnr.

---

#126024 didn't fix the ICE in https://github.com/rust-lang/rust/issues/119272 ([proof](https://rust.godbolt.org/z/GxWK4x9Kj)). Enabling the new solver in coherence did fix both the codegen ICE and incorrectly accepting the code in the first place. 

It seems the theoretical unsoundness fixed in #126024 didn't yet have an MCVE showcasing the issue, and didn't have tests. The correct fix is in principle to use the new solver. 

However, that is not happening soon and until then #126024 has a big impact on compile-times in some cases:
- as seen in [this zulip thread](https://rust-lang.zulipchat.com/#narrow/channel/144729-t-types/topic/Fuchsia.20Build.20Time.20Regression): fuchsia is seeing a 10x regression on the `starnix_core` crate with #126024.
- similarly, as seen in #132064: it looks like one crate is making the entire project build 6-7x slower in 1.82.0. It seemed likely that people would encounter the same behavior when #126024 landed on stable
- even the [post-merge perf results](https://github.com/rust-lang/rust/pull/126024#issuecomment-2248358134) showed significant regressions on a popular crate (`hyper`). Some of the PR's effects should be seen by a large number of people but thankfully they're not always as dramatic as the previous 2 cases.

---

I didn't add the crash test for #119272 back, and the  `auto-traits/opaque_type_candidate_selection.rs` test was deleted in #130654. I assume the latter was because there already was coverage for such a case, but if not I can add that test back, updated for `-Znext-solver=coherence`. <sub>I don't have an MCVE for #132064 but *if needed*, I could try to minimize it.</sub>

Fixes #132064.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11485003342/job/31964102467) [(plain)](https://github.com/rust-lang/rust/commit/9f4b97876616a6f5a8cddd2399fcf2d8ec385a18/checks/31964102467/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
9 
+ error[E0492]: constants cannot refer to interior mutable data
+   --> $DIR/const-promoted-opaque.rs:32:19
+    |
+ LL | const BAZ: &Foo = &FOO;
+    |                   ^^^^ this borrow of an interior mutable value may end up in the final value
10 error[E0716]: temporary value dropped while borrowed
11   --> $DIR/const-promoted-opaque.rs:36:26
12    |

---
To only update this specific test, also pass `--test-args consts/const-promoted-opaque.rs`

error in revision `string`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/consts/const-promoted-opaque.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "string" "--check-cfg" "cfg(FALSE,string,unit,atomic)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-promoted-opaque.string" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-promoted-opaque.string/auxiliary"
--- stderr -------------------------------
error[E0493]: destructor of `helper::Foo` cannot be evaluated at compile-time
##[error]  --> /checkout/tests/ui/consts/const-promoted-opaque.rs:28:26
   |
   |
LL |     let _: &'static _ = &FOO;
   |                          ^^^ the destructor for this type cannot be evaluated in constants
LL |     //[string,atomic]~^ ERROR: destructor of `helper::Foo` cannot be evaluated at compile-time
LL | };
   | - value is dropped here
error[E0492]: constants cannot refer to interior mutable data
##[error]  --> /checkout/tests/ui/consts/const-promoted-opaque.rs:32:19
   |
   |
LL | const BAZ: &Foo = &FOO;
   |                   ^^^^ this borrow of an interior mutable value may end up in the final value
error[E0716]: temporary value dropped while borrowed
##[error]  --> /checkout/tests/ui/consts/const-promoted-opaque.rs:36:26
   |
LL |     let _: &'static _ = &FOO;
LL |     let _: &'static _ = &FOO;
   |            ----------    ^^^ creates a temporary value which is freed while still in use
   |            |
   |            type annotation requires that borrow lasts for `'static`
LL |     //[string,atomic]~^ ERROR: temporary value dropped while borrowed
   | - temporary value is freed at the end of this statement

error: aborting due to 3 previous errors


Some errors have detailed explanations: E0492, E0493, E0716.
For more information about an error, try `rustc --explain E0492`.
------------------------------------------


---- [ui] tests/ui/type-alias-impl-trait/reveal_local.rs stdout ----

error: Error: expected failure status (Some(1)) but received status Some(101).
status: exit status: 101
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/type-alias-impl-trait/reveal_local.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/reveal_local" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type-alias-impl-trait/reveal_local/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error: cannot check whether the hidden type of `reveal_local[9507]::Foo::{opaque#0}` satisfies auto traits
   |
LL |     is_send::<Foo>();
   |               ^^^
   |
   |
   = note: fetching the hidden types of an opaque inside of the defining scope is not supported. You can try moving the opaque type and the item that actually registers a hidden type into a new submodule
note: opaque type is declared here
   |
LL | type Foo = impl Debug;
   |            ^^^^^^^^^^
note: required by a bound in `is_send`
note: required by a bound in `is_send`
  --> /checkout/tests/ui/type-alias-impl-trait/reveal_local.rs:7:15
   |
LL | fn is_send<T: Send>() {}

thread 'rustc' panicked at compiler/rustc_trait_selection/src/traits/select/mod.rs:2375:21:
internal error: entered unreachable code
stack backtrace:
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: <rustc_trait_selection::traits::select::SelectionContext>::confirm_candidate
   4: <rustc_trait_selection::traits::select::SelectionContext>::poly_select
   5: <rustc_trait_selection::traits::fulfill::FulfillProcessor as rustc_data_structures::obligation_forest::ObligationProcessor>::process_obligation
   6: <rustc_data_structures::obligation_forest::ObligationForest<rustc_trait_selection::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection::traits::fulfill::FulfillProcessor>
   7: <rustc_trait_selection::traits::fulfill::FulfillmentContext<rustc_trait_selection::traits::FulfillmentError> as rustc_infer::traits::engine::TraitEngine<rustc_trait_selection::traits::FulfillmentError>>::select_where_possible
   8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_argument_types
   9: <rustc_hir_typeck::fn_ctxt::FnCtxt>::confirm_builtin_call
  10: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  11: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  12: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation
  13: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_block_with_expected
  14: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  15: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  16: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation
  17: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_return_expr
  18: rustc_hir_typeck::check::check_fn
  19: rustc_hir_typeck::typeck
      [... omitted 2 frames ...]
  20: rustc_middle::query::plumbing::query_get_at::<rustc_query_system::query::caches::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 8]>>>
  21: <rustc_hir_analysis::collect::type_of::opaque::TaitConstraintLocator>::check
  22: <rustc_hir_analysis::collect::type_of::opaque::TaitConstraintLocator as rustc_hir::intravisit::Visitor>::visit_item
  23: rustc_hir::intravisit::walk_mod::<rustc_hir_analysis::collect::type_of::opaque::TaitConstraintLocator>
  24: rustc_hir_analysis::collect::type_of::type_of_opaque
      [... omitted 2 frames ...]
  25: rustc_middle::query::plumbing::query_get_at::<rustc_query_system::query::caches::DefIdCache<rustc_middle::query::erase::Erased<[u8; 8]>>>
  26: rustc_hir_analysis::collect::type_of::type_of
      [... omitted 2 frames ...]
  27: rustc_middle::query::plumbing::query_get_at::<rustc_query_system::query::caches::DefIdCache<rustc_middle::query::erase::Erased<[u8; 8]>>>
  29: rustc_hir_analysis::check::wfcheck::check_well_formed
  29: rustc_hir_analysis::check::wfcheck::check_well_formed
      [... omitted 2 frames ...]
  30: rustc_middle::query::plumbing::query_ensure_error_guaranteed::<rustc_query_system::query::caches::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 1]>>, ()>
  31: <rustc_data_structures::sync::parallel::ParallelGuard>::run::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_data_structures::sync::parallel::enabled::try_par_for_each_in<&[rustc_hir::hir::ItemId], rustc_span::ErrorGuaranteed, <rustc_middle::hir::ModuleItems>::par_items<rustc_hir_analysis::check::wfcheck::check_mod_type_wf::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}::{closure#0}>
  32: <rustc_middle::hir::ModuleItems>::par_opaques::<rustc_hir_analysis::check::wfcheck::check_mod_type_wf::{closure#4}>
  33: rustc_hir_analysis::check::wfcheck::check_mod_type_wf
      [... omitted 2 frames ...]
  34: <rustc_middle::hir::map::Map>::par_for_each_module::<rustc_hir_analysis::check_crate::{closure#0}::{closure#0}>::{closure#0}
  35: <rustc_data_structures::sync::parallel::ParallelGuard>::run::<(), rustc_data_structures::sync::parallel::enabled::par_for_each_in<&rustc_hir::hir_id::OwnerId, &[rustc_hir::hir_id::OwnerId], <rustc_middle::hir::map::Map>::par_for_each_module<rustc_hir_analysis::check_crate::{closure#0}::{closure#0}>::{closure#0}>::{closure#0}::{closure#1}::{closure#0}>
  37: rustc_hir_analysis::check_crate
  38: rustc_interface::passes::analysis
  38: rustc_interface::passes::analysis
      [... omitted 2 frames ...]
  40: <rustc_interface::interface::Compiler>::enter::<rustc_driver_impl::run_compiler::{closure#0}::{closure#1}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_span::ErrorGuaranteed>>
  41: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>
  42: rustc_span::create_session_globals_then::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
---
note: please make sure that you have updated to the latest nightly

note: rustc 1.84.0-nightly (a0b76614a 2024-10-23) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0
query stack during panic:
query stack during panic:
#0 [typeck] type-checking `not_gooder`
#1 [type_of_opaque] computing type of opaque `Foo::{opaque#0}`
#2 [type_of] computing type of `Foo::{opaque#0}`
#3 [check_well_formed] checking that `Foo::{opaque#0}` is well-formed
#4 [check_mod_type_wf] checking that types are well-formed in top-level module
end of query stack
error: aborting due to 1 previous error
------------------------------------------

```

</details>

lqd commented:
The first failure is just a test needing expectations updated, but the second is that `type-alias-impl-trait/reveal_local.rs` now ICEs in #122192 code, which is understandable since https://github.com/rust-lang/rust/pull/126024 was extracted from it. 

Just to check, all UI tests pass with #122192 also reverted.

bors commented:
:umbrella: The latest upstream changes (presumably #132349) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors commented:
Hi @lqd, can you rebase this PR and getting this into a landable state? I'm in favor of landing this, especially because the unsoundness it fixes was never really deeply investigated AFAICT, and it's affecting people on stable. cc @lcnr

compiler-errors commented:
@rustbot author

lqd committed:
use global cache when defining opaque types

the compile time impact is too great for a theoretical issue.

This basically reverts commit 61b5e11c47c452ed987295b147783e8bd8e16434

lqd committed:
partially revert acba6449f8597c85b450d96eccda31025cdc0049

This avoids the ICE in `tests/ui/type-alias-impl-trait/reveal_local.rs`.

lqd committed:
update test expectations

lqd commented:
> Hi @lqd, can you rebase this PR and getting this into a landable state?

Hi @compiler-errors,

I can get this into a _reviewable_ state (and just pushed), but I'm not sure I can get it into a landable state per se: this is getting too subtle for my puny brain.

As mentioned in my last comment, some of #122192 depends on #126024, and I've also reverted part of it so that tests pass. This seems somewhat ok as it was just an `unreachable!()` protecting the previous status quo, but I'm not sure of the subtle effects and if it was only exercised via TAITs. 

I've tried to add comments, but we'd likely need actual issues to track this work, even if it would end up being fixed only when the new solver stabilizes.

So that's why it's still in draft. At this point, I'm not sure if I'm just wasting your time and it'd have been faster if you/lcnr had done the revert instead.

@rustbot ready

compiler-errors commented:
Oof, sorry I have not been following closely enough to understand the problem with the revert. I'll take a closer look -- thanks for pushing what you have.

bors commented:
:umbrella: The latest upstream changes (presumably #132625) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
